### ISSUE TITLE
Service should log information about exceptional conditions before raising an exception

### ISSUE THREAD

**Issue by crabhi** [openscoring/openscoring#50]
When multiple clients try to load the same model in parallel, the server may fail with an internal server error.

Unfortunately, Openscoring doesn't log any error. Is it possible to turn on the error logging in Openscoring?

The body of the failure response is:

```json
{
  "message" : "Concurrent modification"
}
```

Testing code:
```python
import logging
import requests

from multiprocessing import Pool

workers = 2  # changing this to 1 fixes the issue
logging.basicConfig(format='%(asctime)-15s %(message)s')
logger = logging.getLogger()
state = 'connection error'


with open('DecisionTreeIris.pmml') as f:
    model = f.read()

def upload_model(n, *args, **kwargs):
    r = requests.put(
        f'http://localhost:8080/openscoring/model/DecisionTreeIris',
        data=model,
        timeout=10,)
    r.raise_for_status()
    return 1

with Pool(workers) as p:
    while True:
        prev_state = state
        try:
            res = p.map(upload_model, range(workers))
            print(res)
        except requests.ConnectionError:
            state = 'connection error'
        else:
            state = 'normal'

        if prev_state != state:
            logger.error(state)

```

Testcode log:
```
$ python3 openscoring_test.py
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/usr/lib/python3.6/multiprocessing/pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "/usr/lib/python3.6/multiprocessing/pool.py", line 44, in mapstar
    return list(map(*args))
  File "openscoring_test.py", line 20, in upload_model
    r.raise_for_status()
  File "/home/krab/.local/lib/python3.6/site-packages/requests/models.py", line 940, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 500 Server Error: Internal Server Error for url: http://localhost:8080/openscoring/model/DecisionTreeIris
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "openscoring_test.py", line 27, in <module>
    res = p.map(upload_model, range(workers))
  File "/usr/lib/python3.6/multiprocessing/pool.py", line 288, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/usr/lib/python3.6/multiprocessing/pool.py", line 670, in get
    raise self._value
requests.exceptions.HTTPError: 500 Server Error: Internal Server Error for url: http://localhost:8080/openscoring/model/DecisionTreeIris
```

Openscoring log:

```
$ java -jar openscoring-server-executable-1.4.5.jar
Jun 13, 2019 11:27:49 AM org.eclipse.jetty.util.log.Log initialized
INFO: Logging initialized @164ms to org.eclipse.jetty.util.log.Slf4jLog
Jun 13, 2019 11:27:49 AM org.openscoring.service.filters.NetworkSecurityContextFilter discoverLocalAddresses
INFO: Local network addresses: [127.0.1.1, 127.0.0.1]
Jun 13, 2019 11:27:49 AM org.openscoring.service.filters.ServiceIdentificationFilter discoverNameAndVersion
INFO: Service name and version: Openscoring/1.4.5
Jun 13, 2019 11:27:49 AM org.eclipse.jetty.server.Server doStart
INFO: jetty-9.4.z-SNAPSHOT; built: 2018-11-14T21:20:31.478Z; git: c4550056e785fb5665914545889f21dc136ad9e6; jvm 11.0.3+7-Ubuntu-1ubuntu218.04.1
Jun 13, 2019 11:27:50 AM org.eclipse.jetty.server.handler.ContextHandler doStart
INFO: Started o.e.j.s.ServletContextHandler@105fece7{/openscoring,null,AVAILABLE}
Jun 13, 2019 11:27:50 AM org.eclipse.jetty.server.AbstractConnector doStart
INFO: Started ServerConnector@40bffbca{HTTP/1.1,[http/1.1]}{0.0.0.0:8080}
Jun 13, 2019 11:27:50 AM org.eclipse.jetty.server.Server doStart
INFO: Started @925ms
Jun 13, 2019 11:27:57 AM org.openscoring.service.filters.NetworkSecurityContext isUserInRole
INFO: Admin role granted to network address 127.0.0.1
Jun 13, 2019 11:27:57 AM org.openscoring.service.filters.NetworkSecurityContext isUserInRole
INFO: Admin role granted to network address 127.0.0.1
```

**Comment by vruusmann** [comment]
> When multiple clients try to load the same model in parallel, the server may fail with an internal server error.

That is as "artificial" issue as it gets - there is no reasonable application scenario where multiple agents might want to update the same model within <1 ms of each other. In reality, a model is updated once per day or once per week, by a human operator that has received an approval from another human to do so.

However, there's a valid point that Openscoring currently does not log all exceptional conditions. For example, if somebody is sending a request to a non-existent model id, then a `ModelNotFoundException` is raised, but it's not recorded in the log file.

**Comment by crabhi** [comment]
Hello, thanks for the quick response. :wave: 

> That is as "artificial" issue as it gets

It's a bug from our production environment, definitely not artificial. We have an application that evaluates models in Openscoring. This application is also the source of the models and it uploads the model into Openscoring when it finds out it's missing. 

The model is deployed much more frequently than once per day because

a) there are multiple servers which need to get restarted on a schedule because of patching. After the restart, the fresh Openscoring needs its models. Actually, we restart Openscoring more frequently for another reason we could avoid but we can't escape the server rotation.
b) The application that communicates with Openscoring is a queue processing worker that runs in parallel in multiple instances.

There are obviously workarounds possible on our side as well as the possibility to patch Openscoring. I just wanted to explain it's a real issue. The lack of logging just made it a bit longer to discover the cause of the error.

**Comment by vruusmann** [comment]
> This application is also the source of the models and it uploads the model into Openscoring when it finds out it's missing.

How does this lead to a situation where the same model identifier is being updated by multiple agents within <1 ms?

Seems more like an issue with your application architecture. Updateable models should be in a queue; there might be multiple worker threads pulling models from the queue, and uploading to openscoring, but these multiple threads should have some higher-level coordination to make sure that they are not trying to work on the same model identifier within <1ms.

How the hell do you know what model is running in production, if you're updating models concurrently, without any supervision?

**Comment by crabhi** [comment]
Multiple clients may execute this sequence of actions in parallel:

1. Try evaluation
2. Find out no model exists under the ID.
3. Upload the model.
4. Evaluate again.

Once a model gets an ID, it will never change. A new version of the model will get a new ID.

Once I know the parallel upload of model is unsupported, I can work around it in the application architecture, no big deal.

**Comment by vruusmann** [comment]
> Once I know the parallel upload of model is unsupported..

Parallel upload of models is supported.

But we need to agree on the definition of "parallel upload" first. I'm talking about a situation where there a N clients/threads uploading N models, each model with a unique identifier.

You're attempting to reuse the same model identifier. This is concurrent upload (the first thread/client to claim a specific model identifier gets to act on it), not parallel upload.

The concurrent upload bit affects a `Model#put(String key, Model model)` method invocation. How much time does Java's `Map#put` take to execute, a fraction of a microsecond? And your application is running into concurrency issues around this method invocation?

Well, you should appreciate that Openscoring contains a guard condition for such a rare concurrent update situation, and fails properly (instead of silently overwriting previous work).

**Comment by crabhi** [comment]
> How much time does Java's Map#put take to execute, a fraction of a microsecond?

I don't know. I just provided a test code replicating the issue reliably with only two clients and no synchronization with Openscoring internals.

> Well, you should appreciate that Openscoring contains a guard condition for such a rare concurrent update situation, and fails properly (instead of silently overwriting previous work).

I appreciate Openscoring for its stability and good usability in general. Raising an error here is a feature I can't find a use case for. Therefore, I raised the issue and tried to explain how I got there. I appreciate your response and I understand your point. Thank you for your work.
